File: epi/PhoneNumberMnemonic.java
package epi;
import epi.test_framework.EpiTest;
import epi.test_framework.EpiTestComparator;
import epi.test_framework.GenericTest;

import java.util.*;
import java.util.function.BiPredicate;
public class PhoneNumberMnemonic {
  private static char[][] map = {{'0'}, {'1'}
          , {'A', 'B', 'C'}, {'D', 'E', 'F'}
          , {'G', 'H', 'I'}, {'J', 'K', 'L'}
          , {'M', 'N', 'O'}, {'P', 'Q', 'R', 'S'}
          , {'T', 'U', 'V'}, {'W', 'X', 'Y', 'Z'}};
  @EpiTest(testDataFile = "phone_number_mnemonic.tsv")

  public static List<String> phoneMnemonic(String phoneNumber) {
    /*List<String> rtList = new ArrayList<>();
    char[] pNumArray = phoneNumber.toCharArray();

    core(pNumArray, 0, rtList);

    return rtList;*/

//    //Iterative
//    List<String> rtList = new ArrayList<>();
//    rtList.add("");
//
//    StringBuilder tmpSb = new StringBuilder();
//    //T O(4^N * N) S O(4^N)
//    for(char nCh : phoneNumber.toCharArray()) {
//      int n = nCh - '0';
//      List<String> tmp = new ArrayList<>();
//      for(char ch : map[n]) {
//        for(String str : rtList) {
//          tmpSb.append(str).append(ch);
//          tmp.add(tmpSb.toString());
//          tmpSb.setLength(0);
//        }
//      }
//      rtList = new ArrayList<>(tmp);
//    }
//
//    return rtList;

    return epi.kt.PhoneNumberMnemonic.INSTANCE.phoneMnemonic(phoneNumber);
  }

  /*
    "2276696"
    c         0
    rtList

    T O(4^N * N) S O(4^N + N)
 */
  private static void core(char[] rts, int c, List<String> rtList) {
    if(c == rts.length) {
      rtList.add(new String(rts));
      return;
    }

    char ori = rts[c];
    int key = (ori - '0');
    for(char val : map[key]) {
      rts[c] = val;
      core(rts, c + 1, rtList);
    }
    rts[c] = ori;
  }

  @EpiTestComparator
  public static boolean comp(List<String> expected, List<String> result) {
    if (result == null) {
      return false;
    }
    Collections.sort(expected);
    Collections.sort(result);
    return expected.equals(result);
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "PhoneNumberMnemonic.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi/kt/PhoneNumberMnemonic.kt
package epi.kt

import java.lang.StringBuilder

object PhoneNumberMnemonic {
    var map = arrayOf(arrayOf('0'), arrayOf('1'), arrayOf('A', 'B', 'C')
            , arrayOf('D', 'E', 'F'), arrayOf('G', 'H', 'I'), arrayOf('J', 'K', 'L')
            , arrayOf('M', 'N', 'O'), arrayOf('P', 'Q', 'R', 'S'), arrayOf('T', 'U', 'V')
            , arrayOf('W', 'X', 'Y', 'Z'))

    fun phoneMnemonic(phoneNumber: String): List<String> {
        /*var chars = phoneNumber.toCharArray()
        var rtList = ArrayList<String>()

        core(chars, 0, rtList)
        return rtList*/

        //Iterative, T O(4^N * 4^N) S O(4^N)
        var rtList = ArrayList<String>()
        rtList.add("")
        for(ch in phoneNumber.toCharArray()) {
            var tmp = ArrayList<String>()
            for(ch in map[ch - '0']) {
                for(cur in rtList) {
                    tmp.add(StringBuilder(cur).append(ch).toString())
                }
            }
            rtList = ArrayList(tmp);
        }

        return rtList
    }

    fun core(chars: CharArray, cur: Int, rtList: MutableList<String>): Unit {
        if(cur == chars.size) {
            rtList.add(String(chars))
            return
        }

        var ori = chars[cur]
        var idx = ori - '0'
        for(ch in map[idx]) {
            chars[cur] = ch
            core(chars, cur + 1, rtList)
        }
        chars[cur] = ori
    }
}